lib/web/command.go

```diff
@@ -20,20 +20,611 @@
+	}
+
+	if c.ConversationID == "" {
+		return trace.BadParameter("missing conversation ID")
+	}
+
+	if c.ExecutionID == "" {
+		return trace.BadParameter("missing execution ID")
+	}
+
+	return nil
+}
+
+// executeCommand executes a command on all nodes that match the query.
 func (h *Handler) executeCommand(
 	w http.ResponseWriter,
 	r *http.Request,
 	_ httprouter.Params,
 	sessionCtx *SessionContext,
 	site reversetunnel.RemoteSite,
 ) (any, error) {
-	return nil, trace.NotImplemented("not implemented")
+	q := r.URL.Query()
+	params := q.Get("params")
+	if params == "" {
+		return nil, trace.BadParameter("missing params")
+	}
+	var req *CommandRequest
+	if err := json.Unmarshal([]byte(params), &req); err != nil {
+		return nil, trace.BadParameter("failed to read JSON message: %v", err)
+	}
+
+	if err := req.Check(); err != nil {
+		return nil, trace.BadParameter("invalid payload: %v", err)
+	}
+
+	clt, err := sessionCtx.GetUserClient(r.Context(), site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	identity, err := createIdentityContext(req.Login, sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, err := h.cfg.SessionControl.AcquireSessionContext(r.Context(), identity, h.cfg.ProxyWebAddr.Addr, r.RemoteAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	authAccessPoint, err := site.CachingAccessPoint()
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to get auth access point.")
+		return nil, trace.Wrap(err)
+	}
+
+	netConfig, err := authAccessPoint.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to fetch cluster networking config.")
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName := site.GetName()
+
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		errMsg := "Error upgrading to websocket"
+		h.log.WithError(err).Error(errMsg)
+		http.Error(w, errMsg, http.StatusInternalServerError)
+		return nil, nil
+	}
+
+	defer func() {
+		ws.WriteMessage(websocket.CloseMessage, nil)
+		ws.Close()
+	}()
+
+	keepAliveInterval := netConfig.GetKeepAliveInterval()
+	err = ws.SetReadDeadline(deadlineForInterval(keepAliveInterval))
+	if err != nil {
+		h.log.WithError(err).Error("Error setting websocket readline")
+		return nil, nil
+	}
+
+	hosts, err := findByQuery(ctx, clt, req.Query)
+	if err != nil {
+		log.WithError(err).Warn("failed to find nodes by labels")
+	}
+
+	if len(hosts) == 0 {
+		const errMsg = "no servers found"
+		h.log.Error(errMsg)
+		return nil, trace.Errorf(errMsg)
+	}
+
+	h.log.Debugf("found %d hosts", len(hosts))
+
+	for _, host := range hosts {
+		err := func() error {
+			sessionData, err := h.generateCommandSession(&host, req.Login, clusterName, sessionCtx.cfg.User)
+			if err != nil {
+				h.log.WithError(err).Debug("Unable to generate new ssh session.")
```

Comment:
Do we need to keep logging every error additionally? Caller should log it, no?

Reply:
The caller does not log it. Our server handler assumes that the caller is an HTTP client, and then the error is returned. Here we've already upgraded the connection to WebSocket, so the client doesn't care about the HTTP responses. 

I can remove the log messages, but then there is no trace in the logs of what went wrong. I think we have two options here:
1. Leave as it is as it matches the implementation of `TerminalHandler`, for example, https://github.com/gravitational/teleport/blob/6db605fbb854d67aea96398b4e12abccdf61e4f8/lib/web/apiserver.go#L2619
2. I can clean it up a bit by:
  * returning error as HTTP response - the client should be able to read those
  * after the WS connection is created, return any error as a WS message. The client should be able to handle those.
  As a side effect, we will have no indication in Teleport logs of what happens, which can be a good and bad thing.

WDYGT @r0mant @zmb3

Reply:
I vote for #1 for now.

There are a few other related issues where we fail to respond with errors after the WS upgrade completes, I think we should fix them all at once and not make you do it now.